program/freedom-metal-program.wake

```diff
@@ -1,131 +1,91 @@
-  def prefix =
-    extract `(.*)\.dts` dts.getPathName
-    | head
-    | getOrElse dts.getPathName
-
-  match dts.read
-    Pass dtsContents =
-      dtsContents
-      | tokenize `\n`
-      | addChosenNode
-      | (omap addStdoutChosenFromAlias stdoutPath | getOrElse (_))
-      | (omap addEntry metalEntryPairOpt | getOrElse (_))
-      | (omap "metal,rom".addChosenRef romNode | getOrElse (_))
-      | (omap "metal,ram".addChosenRef ramNode | getOrElse (_))
-      | catWith "\n"
-      | write "{prefix}_fixed.dts"
-    fail = dts
-
-global def generateMetalFromDts dts machineName outputDir =
+      match `RISCV_ABI`.getAttribute `RISCV_ARCH`.getAttribute `RISCV_CMODEL`.getAttribute
+        (Some mabi) (Some march) (Some cmodel) = Pass (FreedomMakeAttributes attributes march mabi cmodel)
+        _ _ _ = Fail "Generated make attributes file not formatted correctly: {attributes.getPathName}".makeError
+    Fail fail = Fail fail
+
+tuple FreedomBSP =
+  global LinkerScript:   Path
+  global Header:         Path
+  global InlineHeader:   Path
+  global PlatformHeader: Path
+  global Attributes:     FreedomMakeAttributes
+
+global def makeFreedomBSP coreDTS customOverlay type prefix =
+  # Generate the overlay
+  def designDTSName = "{prefix}-design.dts"
+  def designDTS =
+    match customOverlay
+      Some overlay =
+        # Add coreDTS to the include list in the overlay
+        def coreFilename = extract `.*/(.*)` coreDTS.getPathName | head | getOrElse coreDTS.getPathName
+        def list = coreFilename, Nil ++ overlay.getDevicetreeCustomOverlayIncludes
+        overlay
+        | setDevicetreeCustomOverlayIncludes list
+        | (writeDevicetreeCustomOverlay designDTSName _)
+      None =
+        makeDevicetreeOverlayGeneratorOptions coreDTS Nil type designDTSName
+        | runDevicetreeOverlayGenerator
+        | getJobOutput
+
+  # Create the DTB for the header generators
+  def designDTB = dtsTodtb designDTS (coreDTS, Nil) "{prefix}.dtb"
+
+  # Run the header generators
+  def header =
+    makeFreedomMetalHeaderGeneratorOptions designDTB "{prefix}.h"
+    | runFreedomMetalHeaderGenerator
+
+  def platformHeader =
+    makeFreedomBareHeaderGeneratorOptions designDTB "{prefix}-platform.h"
+    | runFreedomBareHeaderGenerator
+
+  # Run linker script generator
+  def linkerScript =
+    makeLdScriptGeneratorOptions designDTS (coreDTS, Nil) LDSCRIPT_DEFAULT "{prefix}.lds"
+    | runLdScriptGenerator
+    | getJobOutput
+
+  # Create the settings for the Makefile
+  def attributesResult =
+    makeESDKSettingsGeneratorOptions designDTS (coreDTS, Nil) type "{prefix}.mk"
+    | runFreedomMakeAttributesGenerator
+
+  match attributesResult
+    Pass attributes =
+      Pass (
+        FreedomBSP
+        linkerScript
+        header.getFreedomMetalHeaderGeneratorOutputsHeader
+        header.getFreedomMetalHeaderGeneratorOutputsInlineHeader
+        platformHeader
+        attributes
+      )
+    Fail err = Fail err
```

Comment:
FYI, if you find yourself doing this pattern of `match` on a `Result` and just passing through the `Fail` case a lot, you can replace it with [`rmap`](https://github.com/sifive/wake/blob/2384a4e456ab7f8104e0030830f7c71f517223de/share/wake/lib/core/result.wake#L40-L42) or [`rmapPass`](https://github.com/sifive/wake/blob/2384a4e456ab7f8104e0030830f7c71f517223de/share/wake/lib/core/result.wake#L44-L48), which are like `map` and `mapFlat` for `Result` types, respectively. In addition to not having to manually write out the failure passthrough, it helps avoid deeply nesting code.

For example, in this case, you could rewrite it as:

```wake
attributesResult
| rmap (\attributes
  FreedomBSP
  linkerScript
  header.getFreedomMetalHeaderGeneratorOutputsHeader
  header.getFreedomMetalHeaderGeneratorOutputsInlineHeader
  platformHeader
  attributes
)
```